File: java/CountingSort.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class CountingSort {
    public static void main(String[] args) throws IOException {
        String nomeArquivo = null; 
        String caminhoArquivoEntrada = System.getProperty("user.dir") + "/../inputs/int_95_ordenado/8_500000_10000000.txt";
        String caminhoArquivoSaida = System.getProperty("user.dir") + "/../output/int_Couting_java";
        try {
            BufferedReader br = new BufferedReader(new FileReader(caminhoArquivoEntrada));
            List<Integer> vetor = new ArrayList<>();
            List<Integer> vetor2 = new ArrayList<>();
            nomeArquivo = br.readLine();
            String line;

            while ((line = br.readLine()) != null) {
                vetor.add(Integer.parseInt(line));
                vetor2.add(0);
            }

            br.close();

            long inicio = System.currentTimeMillis();
            int k = Collections.max(vetor);
            countingSort(vetor,vetor2, k);
            long fim = System.currentTimeMillis();

            //System.out.println(vetor2);

            try (FileWriter fw = new FileWriter(caminhoArquivoSaida, true)) {
                fw.write(nomeArquivo + "\n\tAlgoritmo de ordenacao: CoutingSort\n\tTempo de execucao: " + (fim - inicio) + " ms\n\tQuantidade de numeros: " + vetor.size() + "\n\n");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void countingSort(List<Integer> vetorA, List<Integer> vetorB, int k) {
        List<Integer> vetorC = new ArrayList<>();
        for(int i = 0; i <=k; i++){
            vetorC.add(i,0);
        }
        for(int j = 0; j < vetorA.size(); j++){
            int value = vetorA.get(j);
            vetorC.set(value, vetorC.get(value) + 1);
        }
        for (int i = 1; i <= k; i++){
            vetorC.set(i, vetorC.get(i)+ vetorC.get(i-1));
        }
        for (int j = vetorA.size()-1; j >= 0; j--){
            int value = vetorA.get(j);
            int index = vetorC.get(value) - 1;
            vetorB.set(index, value);
            vetorC.set(value, vetorC.get(value) - 1);
        }
    }
}

File: python/countingSort.py
import sys
import time

def entrada(vetor_a):
    try:
        while True:
            vetor_a.append(int(input())) 
    except EOFError:
        print("Arquivo lido com sucesso!")

def countingSort(vetor_a, vetor_b, x, tam):
    vetor_c = [0] * (x+1)

    for j in range(tam):
        vetor_c[vetor_a[j]] += 1

    for i in range(1, x+1):
        vetor_c[i] += vetor_c[i-1]

    for j in range(tam-1, 0, -1):
        vetor_b[vetor_c[vetor_a[j]]-1] = vetor_a[j]
        vetor_c[vetor_a[j]] -= 1


vetor_a = []
vetor_a_cp = []

#recebe a primeira linha do arquivo
nome_arquivo = input()

entrada(vetor_a)
x = max(vetor_a)
tam = len(vetor_a)


for c in range (3):
    vetor_a_cp = vetor_a.copy()
    vetor_b = [0] * tam 

    inicio = time.perf_counter()
    countingSort(vetor_a_cp, vetor_b, x, tam)
    fim = time.perf_counter()

    arquivo = open('python/output/countingSort.txt', 'a')
    arquivo.write(f"{(fim - inicio) * 1000}\t")

arquivo.write("\n")
arquivo.close()